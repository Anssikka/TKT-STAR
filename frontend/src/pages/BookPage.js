import React, { useState, useEffect } from 'react'
import bookService from '../services/books'
import PageTitle from '../components/PageTitle'
import BookList from '../components/books/BookList'
import AddBookForm from '../components/books/AddBookForm'
import PageSubtitle from '../components/PageSubtitle'
import '../styles/BookPage.css'

function BookPage() {
  const [books, setBooks] = useState([])
  useEffect(() => {
    getBooks()
  }, [])

  const getBooks = async () => {
    try {
      const savedBooks = await bookService.getBooks()
      setBooks(savedBooks)
    } catch (error) {
      console.log(error)
    }
  }

  const addBook = async book => {
    try {
      const createdBook = await bookService.createBook(book)
      setBooks(books.concat(createdBook))
    } catch (error) {
      console.log(error)
    }
  }

  const handleToggle = async id => {
    const book = books.find(b => b.id === id)

    if (book) {
      const updatedBook = await bookService.markAsRead(book.id)
      setBooks(books.map(b => (b.id === updatedBook.id ? updatedBook : b)))
    }
  }

  return (
    <main className="book-page">
      <PageTitle title="Book Recommendations" />
      <PageSubtitle title="Add a Book" />
      <AddBookForm handleSubmit={addBook} />
      <BookList title="Saved Books" books={books} handleToggle={handleToggle} />
    </main>
  )
}

export default BookPage
