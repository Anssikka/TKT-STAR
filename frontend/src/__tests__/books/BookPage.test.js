import React from 'react'
import '@testing-library/jest-dom/extend-expect'
import { render, cleanup, waitForElement } from '@testing-library/react'
import BookPage from '../../pages/BookPage'
import axios from 'axios'

jest.mock('axios')

afterEach(cleanup)

describe('<BookPage />', () => {
  it('has correct elements', async () => {
    axios.get.mockResolvedValue({
      data: [
        {
          id: 1,
          title: 'Test 101',
          author: 'Tester',
          isbn: '1234-4321',
          tags: [
            { id: 1, title: 'Tag #1' },
            { id: 2, title: 'Tag #2' }
          ]
        },
        {
          id: 2,
          title: 'Only Title'
        }
      ]
    })
    const component = render(<BookPage />)
    await waitForElement(() => component.findByText('Test 101'))
    expect(component.asFragment()).toMatchSnapshot()
  })
})
