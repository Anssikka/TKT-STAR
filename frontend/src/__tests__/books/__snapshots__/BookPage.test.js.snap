// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BookPage /> has correct elements 1`] = `
<DocumentFragment>
  <main
    class="book-page"
  >
    <h1
      class="page-title"
    >
      Book Recommendations
    </h1>
    <h3
      class="page-subtitle"
    >
      Add a Book
    </h3>
    <div
      class="book-form-wrapper"
    >
      <form
        class="book-form"
      >
        <label>
          Title:
        </label>
        <input
          data-testid="add-book-title"
          placeholder="Type your book title"
          type="text"
          value=""
        />
        <label>
          Author:
        </label>
        <input
          data-testid="add-book-author"
          placeholder="Type your book author"
          type="text"
          value=""
        />
        <label>
          ISBN:
        </label>
        <input
          data-testid="add-book-isbn"
          placeholder="Type your book isbn"
          type="text"
          value=""
        />
        <input
          class="form-submit"
          data-testid="add-book-submit"
          type="submit"
          value="Submit"
        />
      </form>
    </div>
    <div>
      <h3
        class="page-subtitle"
      >
        Saved Books
      </h3>
      <div
        class="book"
      >
        <h5>
          Test 101
        </h5>
        <h6
          class="book-author"
        >
          Tester
        </h6>
        <small>
          1234-4321
        </small>
        <button
          class="book-toggle toggle button null"
        >
          Mark as read
        </button>
        <h6
          class="book-tag-title"
        >
          Tags:
        </h6>
        <ul>
          <li>
            Tag #1
          </li>
          <li>
            Tag #2
          </li>
        </ul>
      </div>
      <div
        class="book"
      >
        <h5>
          Only Title
        </h5>
        <button
          class="book-toggle toggle button null"
        >
          Mark as read
        </button>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
